pkg/tfshim/sdk-v2/cty.go

```diff
@@ -0,0 +1,20 @@
+package sdkv2
+
+import (
+	"github.com/golang/glog"
+	"github.com/hashicorp/go-cty/cty"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+)
+
+// makeResourceRawConfig converts the decoded Go values in a terraform.ResourceConfig into a cty.Value that is
+// appropriate for Instance{Diff,State}.RawConfig.
+func makeResourceRawConfig(config *terraform.ResourceConfig, resource *schema.Resource) cty.Value {
+	original := schema.HCL2ValueFromConfigValue(config.Raw)
+	coerced, err := resource.CoreConfigSchema().CoerceValue(original)
```

Comment:
Ok so this was what I was missing!!!

Reply:
Yeah, it took my distressingly long to find that. I implemented the equivalent twice before noticing it.